reanalysis/query_panelapp.py

```diff
@@ -0,0 +1,233 @@
+
+Optionally user can provide a panel version number in the past
+Pull all details from the earlier version
+Store all discrepancies between earlier and current
+
+Write all output to a JSON dictionary
+"""
+
+
+from typing import Any, Dict, List, Optional, Union, Set
+import logging
+import json
+import requests
+
+from cloudpathlib import AnyPath
+
+import click
+
+
+def parse_gene_list(path_to_list: str) -> Set[str]:
+    """
+    parses a file (GCP or local), extracting a set of genes
+    required format: clean data, one per line, gene name only
+    :param path_to_list:
+    """
+    return {
+        line.rstrip()
+        for line in open(AnyPath(path_to_list), 'r', encoding='utf-8')
+        if line.rstrip() != ''
+    }
+
+
+def get_json_response(url: str) -> Union[List[Dict[str, str]], Dict[str, Any]]:
+    """
+    takes a request URL, checks for healthy response, returns the JSON
+    :param url:
+    :return:
+    """
+
+    response = requests.get(url, headers={'Accept': 'application/json'})
+    response.raise_for_status()
+    return response.json()
+
+
+def get_panel_green(
+    panel_id: str = '137',
+    version: Optional[str] = None,
+) -> Dict[str, Dict[str, Union[str, bool]]]:
+    """
+    Takes a panel number, and pulls all GRCh38 gene details from PanelApp
+    For each gene, keep the MOI, symbol, ENSG (where present)
+
+    :param panel_id: defaults to the PanelAppAU Mendeliome
+    :param version:
+    """
+
+    # prepare the query URL
+    panel_app_genes_url = f'https://panelapp.agha.umccr.org/api/v1/{panel_id}'
+    if version is not None:
+        panel_app_genes_url += f'?version={version}'
+
+    panel_response = requests.get(panel_app_genes_url)
+    panel_response.raise_for_status()
+    panel_json = panel_response.json()
+
+    panel_version = panel_json.get('version')
+
+    # pop the version in logging if not manually defined
+    if version is None:
+        logging.info('Current panel version: %s', panel_version)
+
+    gene_dict = {'panel_metadata': {'current_version': panel_version}}
+
+    for gene in panel_json['genes']:
+
+        # only retain green genes
+        if gene['confidence_level'] != '3' or gene['entity_type'] != 'gene':
+            continue
+
+        ensg = None
+        symbol = gene.get('entity_name')
+
+        # take the PanelApp MOI, don't simplify
+        moi = gene.get('mode_of_inheritance', None)
+
+        # for some reason the build is capitalised oddly in panelapp
+        # at least one entry doesn't have an ENSG annotation
+        for build, content in gene['gene_data']['ensembl_genes'].items():
+            if build.lower() == 'grch38':
+                # the ensembl version may alter over time, but will be singular
+                ensg = content[list(content.keys())[0]]['ensembl_id']
+
+        # this appears to be missing in latest panel version
+        if symbol == 'RNU12' and ensg is None:
+            ensg = 'ENSG00000276027'
```

Comment:
Do we want such specific panel fixes here?

Reply:
ideally no, I've reviewed the panel data and found a few curation fixes in order to correctly map panelApp entities to VEP annotations. I'll remove this statement, and in future we'll be reliant on PanelApp as a source of truth